README.md
Business-RO-TaxDeduction
========================
Ștefan Suciu
2016-06-05

Version: 0.010

A Romanian salary tax deduction calculator.

Starting with the v0.004 version, there is a new optional `year`
parameter that can be used to choose between the current regulations
(OMFP 52/2016), or the previous - OMFP 1016/2005.

This is an alternative to the database driven implementation for the
tax deductions calculation.  It may be suitable for small programs or
even for oneliners line this:

```
$ perl -MBusiness::RO::TaxDeduction -E'$td=Business::RO::TaxDeduction->new(vbl=>1400);say $td->tax_deduction'
300
```

It's a little too long but it works ;)


Changes
Revision history for Business-RO-TaxDeduction

0.011   2017-08-15
        Bugfix release.
        Fix for "Invalid coerce '1'".

0.010   2016-06-05
        Add year attribute, now it can make calculations for 2005 -:-
        2015 using the formulas from OMFP 1016/2005 and for 2016 using
        formular from OMFP 52/2016.

0.007   2016-04-21
        Update for 2016 - OMFP52/2016

0.006   2015-10-28
        Refactor round to tens method, add PODs.

0.005   2015-06-13
        Build with dzil.

0.002   2015-06-05
        First version, released on an unsuspecting world.


lib/Business/RO/TaxDeduction/Table.pm
package Business::RO::TaxDeduction::Table;

# ABSTRACT: Deduction table by year

use 5.010001;
use utf8;
use Moo;
use MooX::HandlesVia;
use Business::RO::TaxDeduction::Types qw(
    Int
    HashRef
    TaxPersons
);

with qw(Business::RO::TaxDeduction::Role::Utils);

has vbl => (
    is       => 'ro',
    isa      => Int,
    required => 1,
);

has '_deduction_table_2018_0' => (
    is      => 'ro',
    isa     => HashRef,
    lazy    => 1,
    default => sub {
        return {
            510 => [ 1,    1950 ],
            495 => [ 1951, 2000 ],
            480 => [ 2001, 2050 ],
            465 => [ 2051, 2100 ],
            450 => [ 2101, 2150 ],
            435 => [ 2151, 2200 ],
            420 => [ 2201, 2250 ],
            405 => [ 2251, 2300 ],
            390 => [ 2301, 2350 ],
            375 => [ 2351, 2400 ],
            360 => [ 2401, 2450 ],
            345 => [ 2451, 2500 ],
            330 => [ 2501, 2550 ],
            315 => [ 2551, 2600 ],
            300 => [ 2601, 2650 ],
            285 => [ 2651, 2700 ],
            270 => [ 2701, 2750 ],
            255 => [ 2751, 2800 ],
            240 => [ 2801, 2850 ],
            225 => [ 2851, 2900 ],
            210 => [ 2901, 2950 ],
            195 => [ 2951, 3000 ],
            180 => [ 3001, 3050 ],
            165 => [ 3051, 3100 ],
            150 => [ 3101, 3150 ],
            135 => [ 3151, 3200 ],
            120 => [ 3201, 3250 ],
            105 => [ 3251, 3300 ],
            90  => [ 3301, 3350 ],
            75  => [ 3351, 3400 ],
            60  => [ 3401, 3450 ],
            45  => [ 3451, 3500 ],
            30  => [ 3501, 3550 ],
            15  => [ 3551, 3600 ],
            0   => [ 3601, 999999 ],
        };
    },
);

has '_deduction_table_2018_1' => (
    is      => 'ro',
    isa     => HashRef,
    lazy    => 1,
    default => sub {
        return {
            670 => [ 1,    1950 ],
            655 => [ 1951, 2000 ],
            640 => [ 2001, 2050 ],
            625 => [ 2051, 2100 ],
            610 => [ 2101, 2150 ],
            595 => [ 2151, 2200 ],
            580 => [ 2201, 2250 ],
            565 => [ 2251, 2300 ],
            550 => [ 2301, 2350 ],
            535 => [ 2351, 2400 ],
            520 => [ 2401, 2450 ],
            505 => [ 2451, 2500 ],
            490 => [ 2501, 2550 ],
            475 => [ 2551, 2600 ],
            460 => [ 2601, 2650 ],
            445 => [ 2651, 2700 ],
            430 => [ 2701, 2750 ],
            415 => [ 2751, 2800 ],
            400 => [ 2801, 2850 ],
            385 => [ 2851, 2900 ],
            370 => [ 2901, 2950 ],
            355 => [ 2951, 3000 ],
            340 => [ 3001, 3050 ],
            325 => [ 3051, 3100 ],
            310 => [ 3101, 3150 ],
            295 => [ 3151, 3200 ],
            280 => [ 3201, 3250 ],
            265 => [ 3251, 3300 ],
            250 => [ 3301, 3350 ],
            235 => [ 3351, 3400 ],
            220 => [ 3401, 3450 ],
            205 => [ 3451, 3500 ],
            190 => [ 3501, 3550 ],
            175 => [ 3551, 3600 ],
            0   => [ 3601, 999999 ],
        };
    },
);

has '_deduction_table_2018_2' => (
    is      => 'ro',
    isa     => HashRef,
    lazy    => 1,
    default => sub {
        return {
            830 => [ 1,    1950 ],
            815 => [ 1951, 2000 ],
            800 => [ 2001, 2050 ],
            785 => [ 2051, 2100 ],
            770 => [ 2101, 2150 ],
            755 => [ 2151, 2200 ],
            740 => [ 2201, 2250 ],
            725 => [ 2251, 2300 ],
            710 => [ 2301, 2350 ],
            695 => [ 2351, 2400 ],
            680 => [ 2401, 2450 ],
            665 => [ 2451, 2500 ],
            650 => [ 2501, 2550 ],
            635 => [ 2551, 2600 ],
            620 => [ 2601, 2650 ],
            605 => [ 2651, 2700 ],
            590 => [ 2701, 2750 ],
            575 => [ 2751, 2800 ],
            560 => [ 2801, 2850 ],
            545 => [ 2851, 2900 ],
            530 => [ 2901, 2950 ],
            515 => [ 2951, 3000 ],
            500 => [ 3001, 3050 ],
            485 => [ 3051, 3100 ],
            470 => [ 3101, 3150 ],
            455 => [ 3151, 3200 ],
            440 => [ 3201, 3250 ],
            425 => [ 3251, 3300 ],
            410 => [ 3301, 3350 ],
            395 => [ 3351, 3400 ],
            380 => [ 3401, 3450 ],
            365 => [ 3451, 3500 ],
            350 => [ 3501, 3550 ],
            335 => [ 3551, 3600 ],
            0   => [ 3601, 999999 ],
        };
    },
);

has '_deduction_table_2018_3' => (
    is      => 'ro',
    isa     => HashRef,
    lazy    => 1,
    default => sub {
        return {
            990 => [ 1,    1950 ],
            975 => [ 1951, 2000 ],
            960 => [ 2001, 2050 ],
            945 => [ 2051, 2100 ],
            930 => [ 2101, 2150 ],
            915 => [ 2151, 2200 ],
            900 => [ 2201, 2250 ],
            885 => [ 2251, 2300 ],
            870 => [ 2301, 2350 ],
            855 => [ 2351, 2400 ],
            840 => [ 2401, 2450 ],
            825 => [ 2451, 2500 ],
            810 => [ 2501, 2550 ],
            795 => [ 2551, 2600 ],
            780 => [ 2601, 2650 ],
            765 => [ 2651, 2700 ],
            750 => [ 2701, 2750 ],
            735 => [ 2751, 2800 ],
            720 => [ 2801, 2850 ],
            705 => [ 2851, 2900 ],
            690 => [ 2901, 2950 ],
            675 => [ 2951, 3000 ],
            660 => [ 3001, 3050 ],
            645 => [ 3051, 3100 ],
            630 => [ 3101, 3150 ],
            615 => [ 3151, 3200 ],
            600 => [ 3201, 3250 ],
            585 => [ 3251, 3300 ],
            570 => [ 3301, 3350 ],
            555 => [ 3351, 3400 ],
            540 => [ 3401, 3450 ],
            525 => [ 3451, 3500 ],
            510 => [ 3501, 3550 ],
            495 => [ 3551, 3600 ],
            0   => [ 3601, 999999 ],
        };
    },
);

has '_deduction_table_2018_4' => (
    is      => 'ro',
    isa     => HashRef,
    lazy    => 1,
    default => sub {
        return {
            1310 => [ 1,    1950 ],
            1295 => [ 1951, 2000 ],
            1280 => [ 2001, 2050 ],
            1265 => [ 2051, 2100 ],
            1250 => [ 2101, 2150 ],
            1235 => [ 2151, 2200 ],
            1220 => [ 2201, 2250 ],
            1205 => [ 2251, 2300 ],
            1190 => [ 2301, 2350 ],
            1175 => [ 2351, 2400 ],
            1160 => [ 2401, 2450 ],
            1145 => [ 2451, 2500 ],
            1130 => [ 2501, 2550 ],
            1115 => [ 2551, 2600 ],
            1100 => [ 2601, 2650 ],
            1085 => [ 2651, 2700 ],
            1070 => [ 2701, 2750 ],
            1055 => [ 2751, 2800 ],
            1040 => [ 2801, 2850 ],
            1025 => [ 2851, 2900 ],
            1010 => [ 2901, 2950 ],
            995  => [ 2951, 3000 ],
            980  => [ 3001, 3050 ],
            965  => [ 3051, 3100 ],
            950  => [ 3101, 3150 ],
            935  => [ 3151, 3200 ],
            920  => [ 3201, 3250 ],
            905  => [ 3251, 3300 ],
            890  => [ 3301, 3350 ],
            875  => [ 3351, 3400 ],
            860  => [ 3401, 3450 ],
            845  => [ 3451, 3500 ],
            830  => [ 3501, 3550 ],
            815  => [ 3551, 3600 ],
            0    => [ 3601, 999999 ],
        };
    },
);

has '_deduction_map' => (
    is          => 'ro',
    handles_via => 'Hash',
    lazy        => 1,
    default     => sub {
        my $self = shift;
        my $year = $self->base_year;
        my $pers = $self->persons;
        my $meth = "_deduction_table_${year}_${pers}";
        return $self->$meth;
    },
    handles => {
        _get_amounts => 'keys',
        _get_range   => 'get',
    },
);

sub deduction {
    my $self = shift;
    my $vbl  = $self->vbl;
    foreach my $amount ( $self->_get_amounts ) {
        my $range = $self->_get_range($amount);
        if (   ( $vbl >= $range->[0] )
            && ( $vbl <= $range->[1] ) )
        {
            return $amount;
        }
    }
    my $msg = "vbl=$vbl, persons=" . $self->persons;
    die "Something went wrong for $msg";
}

1;

__END__

=encoding utf8

=head1 SYNOPSIS

    use Business::RO::TaxDeduction::Table;

    my $tax = Business::RO::TaxDeduction::Table->new(
        year    => 2018,
        persons => 1,
        vbl     => 1950,
    );
    my $amount = $tax->deduction;

=head1 DESCRIPTION

Data module.  This module holds data used for calculating the tax
deduction for the changes made for the year 2018;

=head1 INTERFACE

=head2 ATTRIBUTES

=head3 vbl

=head3 _deduction_table_2018_0

=head3 _deduction_table_2018_1

=head3 _deduction_table_2018_2

=head3 _deduction_table_2018_3

=head3 _deduction_table_2018_4

=head3 _deduction_map

=head2 INSTANCE METHODS

=head3 deduction

=cut


lib/Business/RO/TaxDeduction/Types.pm
package Business::RO::TaxDeduction::Types;

# ABSTRACT: Types for the TaxDeduction module

use 5.010001;
use strict;
use warnings;
use utf8;
use Type::Library 0.040 -base, -declare => qw(
    TaxPersons
    MathBigFloat
);

use Type::Utils -all;
use Types::Standard -types;

BEGIN { extends "Types::Standard" };

declare "TaxPersons",
    as "Int",
    where { $_ >= 0 && $_ <= 4 };

class_type MathBigFloat, { class => 'Math::BigFloat' };

1;

__END__

=head1 SYNOPSIS

  use Business::RO::TaxDeduction::Types qw(
      Int
      MathBigFloat
  );

=over

=item C<< MathBigFloat >>

A L<Math::BigFloat> object instance.

=item C<< TaxPersons >>

A custom type for the number of persons.  Valid values are from 0
(zero) to 4.  If the number is greater than 4, the attribute value is
coerced to 4.

=back


lib/Business/RO/TaxDeduction/Amount.pm
package Business::RO::TaxDeduction::Amount;

# ABSTRACT: Personal deduction amount by year and number of persons

use 5.010001;
use utf8;
use Moo;
use MooX::HandlesVia;
use Business::RO::TaxDeduction::Types qw(
    Int
    HashRef
);
with qw(Business::RO::TaxDeduction::Role::Utils);

has '_deduction_map_2005' => (
    is          => 'ro',
    isa         => HashRef,
    lazy        => 1,
    default     => sub {
        return {
            0 => 250,
            1 => 350,
            2 => 450,
            3 => 550,
            4 => 650,
        };
    },
);

has '_deduction_map_2016' => (
    is          => 'ro',
    isa         => HashRef,
    lazy        => 1,
    default     => sub {
        return {
            0 => 300,
            1 => 400,
            2 => 500,
            3 => 600,
            4 => 800,
        };
    },
);

has '_deduction_map_2018' => (
    is          => 'ro',
    isa         => HashRef,
    lazy        => 1,
    default     => sub {
        return {
            0 => 510,
            1 => 670,
            2 => 830,
            3 => 990,
            4 => 1310,
        };
    },
);

has '_deduction_map' => (
    is          => 'ro',
    handles_via => 'Hash',
    lazy        => 1,
    default     => sub {
        my $self = shift;
        my $year = $self->base_year;
        my $meth = "_deduction_map_$year";
        return $self->$meth;
    },
    handles => {
        _get_deduction_for => 'get',
    }
);

has amount => (
    is      => 'ro',
    isa     => Int,
    lazy    => 1,
    default => sub {
        my $self = shift;
        return $self->_get_deduction_for( $self->persons );
    },
);

1;

__END__

=encoding utf8

=head1 SYNOPSIS

    my $ded = Business::RO::TaxDeduction::Amount->new(
        persons => 4,
        year    => 2018,
    );
    say $ded->amount;

=head1 DESCRIPTION

Data module.  Personal deduction amount by year and number of persons.

=head1 INTERFACE

=head2 ATTRIBUTES

=head3 _deduction_map_2005

Uses the amounts described in the document:

"ORDINUL nr. 1.016/2005 din 18 iulie 2005 privind aprobarea deducerilor
personale lunare pentru contribuabilii care realizează venituri din
salarii la funcția de bază, începând cu luna iulie 2005, potrivit
prevederilor Legii nr. 571/2003 privind Codul fiscal și ale Legii
nr. 348/2004 privind denominarea monedei naționale".

=head3 _deduction_map_2016

Uses the amounts described in the document:

"ORDIN Nr. 52/2016 din 14 ianuarie 2016 privind aprobarea
calculatorului pentru determinarea deducerilor personale lunare pentru
contribuabilii care realizează venituri din salarii la funcţia de
bază, începând cu luna ianuarie 2016, potrivit prevederilor art. 77
alin. (2) şi ale art. 66 din Legea nr. 227/2015 privind Codul fiscal".

=head3 _deduction_map_2018

Uses the amounts described in the document:

ORDONANȚĂ DE URGENȚĂ Nr. 79/2017 din 8 noiembrie 2017 pentru
modificarea și completarea Legii nr. 227/2015 privind Codul fiscal
EMITENT: GUVERNUL ROMÂNIEI PUBLICATĂ ÎN: MONITORUL OFICIAL NR. 885 din
10 noiembrie 2017

=head3 _deduction_map

Choses the appropriate deduction map by year and returns the amount
for the number of persons given as parameter.

=head3 amount

Returns the C<amount> using the C<_deduction_map> method.

=head2 INSTANCE METHODS

=cut


lib/Business/RO/TaxDeduction/Ranges.pm
package Business::RO::TaxDeduction::Ranges;

# ABSTRACT: Deduction ranges by year

use 5.010001;
use utf8;
use Moo;
use Business::RO::TaxDeduction::Types qw(
    Int
);
with qw(Business::RO::TaxDeduction::Role::Utils);

has 'vbl_min' => (
    is      => 'ro',
    isa     => Int,
    lazy    => 1,
    default => sub {
        my $self = shift;
        return 1000 if $self->base_year == 2005;
        return 1500 if $self->base_year == 2016;
        return 1950 if $self->base_year == 2018;
        die "Not a valid year: " . $self->base_year;
    },
);

has 'vbl_max' => (
    is      => 'ro',
    isa     => Int,
    lazy    => 1,
    default => sub {
        my $self = shift;
        return 3000 if $self->base_year == 2005;
        return 3000 if $self->base_year == 2016;
        return 3600 if $self->base_year == 2018;
        die "Not a valid year: " . $self->base_year;
    },
);

has 'f_min' => (
    is      => 'ro',
    isa     => Int,
    lazy    => 1,
    default => sub {
        my $self = shift;
        return 1000 if $self->base_year == 2005;
        return 1500 if $self->base_year == 2016;
        return 1950 if $self->base_year == 2018;
        die "Not a valid year: " . $self->base_year;
    },
);

has 'f_max' => (
    is       => 'ro',
    isa      => Int,
    lazy    => 1,
    default => sub {
        my $self = shift;
        return 2000 if $self->base_year == 2005;
        return 1500 if $self->base_year == 2016;
        return 1700 if $self->base_year == 2018;
        die "Not a valid year: " . $self->base_year;
    },
);

1;

__END__

=encoding utf8

=head1 SYNOPSIS

    use Business::RO::TaxDeduction::Ranges;

    my $tdr = Business::RO::TaxDeduction::Ranges->new(
        year => $year,
    );
    my $vbl_min = $tdr->vbl_min;
    my $vbl_max = $tdr->vbl_max;
    my $f_min   = $tdr->f_min;
    my $f_max   = $tdr->f_max;

=head1 DESCRIPTION

Data module.  This module holds data used in the tax deduction
formula.

NOTE: No tax deduction formula for 2018.

=head1 INTERFACE

=head2 ATTRIBUTES

=head3 vbl_min

The minimum amount (VBL) used in the tax deduction formula.

=head3 vbl_max

The maximum amount (VBL) used in the tax deduction formula.

=head3 f_min

The amount used in the tax deduction formula:

    250 x [ 1 - ( VBL - $f_min=1000 ) / 2000 ]  # 2005
    300 x [ 1 - ( VBL - $f_min=1500 ) / 1500 ]  # 2016

=head3 f_max

The amount used in the tax deduction formula:

    250 x [ 1 - ( VBL - 1000 ) / $f_max=2000 ]  # 2005
    300 x [ 1 - ( VBL - 1500 ) / $f_max=1500 ]  # 2016

=head2 INSTANCE METHODS

=cut


lib/Business/RO/TaxDeduction/Role/Utils.pm
package Business::RO::TaxDeduction::Role::Utils;

# ABSTRACT: A utility role

use Moo::Role;
use Business::RO::TaxDeduction::Types qw(
    Int
    TaxPersons
);

has 'persons' => (
    is       => 'ro',
    isa      => TaxPersons,
    required => 1,
    coerce   => sub {
        $_[0] >= 4 ? 4 : $_[0];
    },
    default  => sub { 0 },
);

has 'year' => (
    is       => 'ro',
    isa      => Int,
    default  => sub { 2018 },
);

has 'base_year' => (
    is       => 'ro',
    isa      => Int,
    lazy     => 1,
    default  => sub {
        my $self = shift;
        if ($self->year >= 2018) {
            return 2018;
        }
        if ($self->year >= 2016) {
            return 2016;
        }
        elsif ($self->year >= 2005) {
            return 2005;
        }
        else {
            die "The tax deduction does not apply before 2005!";
        }
    },
);

1;

__END__

=encoding utf8

=head1 SYNOPSIS

    package Business::RO::TaxDeduction::Ranges;
    with qw(Business::RO::TaxDeduction::Role::Utils);

=head1 DESCRIPTION

A role that incapsulates common attributes required by the
TaxDeduction::Amount and TaxDeduction::Ranges modules.

=head1 INTERFACE

=head2 ATTRIBUTES

=head3 persons

Number of persons.  Coerced to 4 if is greater than 4.

=head3 year

The C<year> attribute holds the year of the tax deduction calculation.

=head3 base_year

The C<base_year> attribute represents the year when the methodology of
the tax deduction calculation was introduced.  Is calculated using the
C<year> attribute and currently can take one of the two values: 2005
or 2016.

=head2 INSTANCE METHODS

=cut


lib/Business/RO/TaxDeduction.pm
package Business::RO::TaxDeduction;

# ABSTRACT: Romanian salary tax deduction calculator

use 5.010001;
use utf8;
use Moo;
use MooX::HandlesVia;
use Math::BigFloat;
use Scalar::Util qw(blessed);
use Business::RO::TaxDeduction::Types qw(
    Int
    MathBigFloat
);
use Business::RO::TaxDeduction::Amount;
use Business::RO::TaxDeduction::Ranges;
use Business::RO::TaxDeduction::Table;

has 'vbl' => (
    is       => 'ro',
    isa      => Int,
    required => 1,
);

has 'year' => (
    is       => 'ro',
    isa      => Int,
    default  => sub { 2018 },
);

has 'persons' => (
    is       => 'ro',
    isa      => Int,
    default  => sub { 0 },
);

has 'deduction' => (
    is      => 'ro',
    lazy    => 1,
    default => sub {
        my $self = shift;
        return Business::RO::TaxDeduction::Amount->new(
            year    => $self->year,
            persons => $self->persons,
        );
    },
);

has 'ranges' => (
    is      => 'ro',
    lazy    => 1,
    default => sub {
        my $self = shift;
        return Business::RO::TaxDeduction::Ranges->new(
            year => $self->year,
        );
    },
    handles => [ qw( vbl_min vbl_max f_min f_max ) ],
);

has 'ten' => (
    is      => 'ro',
    isa     => MathBigFloat,
    default => sub {
        return Math::BigFloat->new(10);
    },
);

sub tax_deduction {
    my $self   = shift;
    my $vbl    = $self->_round_to_int( $self->vbl );

    return $self->_amount_for_2018 if $self->year >= 2018;

    my $amount = $self->deduction->amount;
    if ( $vbl <= $self->vbl_min ) {
        return $amount;
    }
    elsif ( ( $vbl > $self->vbl_min ) && ( $vbl <= $self->vbl_max ) ) {
        $amount = $self->_tax_deduction_formula($vbl, $amount);
        return ( blessed $amount ) ? $amount->bstr : $amount;
    }
    else {
        return 0;               # 0 for VBL > vbl_max
    }
}

sub _tax_deduction_formula {
    my ( $self, $vbl, $base_deduction ) = @_;
    my $amount = $base_deduction * ( 1 - ( $vbl - $self->f_min ) / $self->f_max );
    return $self->_round_to_tens($amount);
}

sub _round_to_int {
    my ( $self, $amount ) = @_;
    return int( $amount + 0.5 * ( $amount <=> 0 ) );
}

sub _round_to_tens {
    my ( $self, $para_amount ) = @_;
    my $amount = Math::BigFloat->new($para_amount);

    return 0 if $amount == 0;

    my $afloor  = $amount->copy()->bfloor();
    my $amodulo = $afloor->copy()->bmod( $self->ten );

    return $amount if $amount->is_int && $amodulo == 0;
    return $afloor->bsub($amodulo)->badd( $self->ten );
}

sub _amount_for_2018 {
    my $self = shift;
    my $table =  Business::RO::TaxDeduction::Table->new(
        year    => $self->year,
        persons => $self->persons,
        vbl     => $self->vbl,
    );
    return $table->deduction;
}

1;

__END__

=encoding utf8

=head1 SYNOPSIS

    use Business::RO::TaxDeduction;

    my $brtd = Business::RO::TaxDeduction->new(
        vbl     => 1400,
        persons => 3,
    );
    my $amount = $brtd->tax_deduction;

=head1 DESCRIPTION

Romanian salary tax deduction calculator.

=head1 INTERFACE

=head2 ATTRIBUTES

=head3 vbl

The C<vbl> attribute holds the input amount of the tax deduction
calculation.  (ro: Venit Brut Lunar).

=head3 year

The C<year> attribute holds the year of the tax deduction calculation.

=head3 persons

The C<persons> attribute holds the number of persons.  Not required,
the default is 0.

=head3 deduction

The C<deduction> attribute holds a
C<Business::RO::TaxDeduction::Amount> object instance.

=head3 ten

A Math::BigFloat object instance for 10.

=head3 five

A Math::BigFloat object instance for 5.

=head2 INSTANCE METHODS

=head3 tax_deduction

Return the deduction calculated for the given amount.

Starting with the current version (0.004) the appropriate algorithm
for the tax deduction calculation year is chosen.

=head3 _tax_deduction_formula

Formula for calculating the tax deduction for amounts above C<vbl_min>
and less or equal to C<vbl_max>.

=head3 _round_to_int

Custom rounding method to the nearest integer.  It uses the Romanian
standard for rounding in bookkeeping.

Example:

  10.01 -:- 10.49 => 10
  10.50 -:- 10.99 => 11

=head3 _round_to_tens

Round up to tens.  Uses Math::BigFloat to prevent rounding errors like
when amount minus floor(amount) gives something like 7.105427357601e-15.

=cut
